### ISSUE TITLE
Configuring the queue to be processed by a separate server to the main application server adds the job to the wrong redis store

### ISSUE THREAD

**Issue by matt-ostrom-hall-rss** [laravel/horizon#1106]
- Horizon Version: 5.7.17
- Laravel Version: 8.76.2
- PHP Version: 7.4.24
- Redis Driver & Version: predis 1.1.9
- Database Driver & Version: MySQL 5.7.36

### Description:

If you configure a custom redis connection for an external server in config/database.php and then set that as the redis queue connection in config/queue.php, queued jobs are added to the local redis store on the application server instead of being pushed to the external server redis store that is configured for use.

### Steps To Reproduce:

- Install a fresh laravel application and horizon.
- Add a new entry to the redis array in config/database.php and add the environment variables for the external redis server
```
'queue' => [
    'url' => env('REDIS_QUEUE_URL'),
    'host' => env('REDIS_QUEUE_HOST', '127.0.0.1'),
    'password' => env('REDIS_QUEUE_PASSWORD', null),
    'port' => env('REDIS_QUEUE_PORT', '6379'),
    'database' => env('REDIS_QUEUE_DB', '0'),
]
```

- Update the connection value to 'queue' for the redis entry in config/queue.php
```
'redis' => [
    'driver' => 'redis',
    'connection' => 'queue',
    'queue' => env('REDIS_QUEUE', 'default'),
    'retry_after' => 90,
    'block_for' => null,
    'after_commit' => false,
]
```
- Create a job and dispatch it
- The job will be a pending job on the server the job was dispatched from and not be in the redis store on the external server.
- https://github.com/Really-Simple-Systems/horizon-bug-report

**Comment by driesvints** [comment]
Are you sure all your env variables are set correctly? Because the Redis connections are identical. Is `REDIS_QUEUE_HOST` pointing to a different url than `REDIS_HOST`?

**Comment by matt-ostrom-hall-rss** [comment]
Yeah the `REDIS_QUEUE_HOST` variable is pointing to the external redis server, while `REDIS_HOST` is pointing to the application server.

**Comment by driesvints** [comment]
I have to be frank: this sounds like a configuration issue as I know of specific apps that are using this properly. I'm fairly sure there's something incorrectly configured on your end.

**Comment by matt-ostrom-hall-rss** [comment]
This happens with the fresh repo I've linked above, the only configuration that's been done is changing the queue connection as above, without horizon installed the job gets put on the redis store on the external server. What could be configured incorrectly?

**Comment by driesvints** [comment]
I wonder if it has to do with these settings for the supervisor: https://github.com/Really-Simple-Systems/horizon-bug-report/blob/main/config/horizon.php#L169-L170

Also: what's the value of `QUEUE_CONNECTION` for you?

**Comment by matt-ostrom-hall-rss** [comment]
I've got it set as `QUEUE_CONNECTION=redis`

**Comment by driesvints** [comment]
I'm out of options I'm afraid. I don't know why this is happening. Maybe one of my team mates knows but most of us are with vacation until the beginning of the year. I'll leave this open for a bit longer but I'm afraid I can't help you further.

**Comment by matt-ostrom-hall-rss** [comment]
Ok thank you for your help, i'll check back in the new year. Have a good holidays :)

**Comment by driesvints** [comment]
@jmarcher I'm gonna have a more thorough look at this soon hopefully. In the meantime can you maybe prepare a repo for me to try out with:

```bash
laravel new bug-report --github="--public"
```

Please make sure that you have the [latest version of the Laravel installer](https://github.com/laravel/installer) in order to run this command. Please also make sure you have both Git & [the GitHub CLI tool](https://cli.github.com) properly set up. Please do not amend and create a separate commit with your custom changes. After you've posted the repository, we'll try to reproduce the issue.

And also list the ENV variables that I should set and prep them a bit some values you're using. I'll then try to recreate the problem myself.

Thanks!

**Comment by mattostromhall** [comment]
@driesvints was this comment intended for this issue? I've created a repo to help reproduce here https://github.com/Really-Simple-Systems/horizon-bug-report

**Comment by matt-ostrom-hall-rss** [comment]
Yeah sure, only bits you need to set specifically for you is DB_DATABASE and REDIS_QUEUE_HOST I think. Thank you :)

```
APP_NAME="Horizon Test"
APP_ENV=local
APP_KEY=base64:0RcS4lKkogxrZSL/flqlDO6TXsKwzfRUjE7i/wfyCvU=
APP_DEBUG=true
APP_URL=http://HorizonTest.test

LOG_CHANNEL=stack
LOG_DEPRECATIONS_CHANNEL=null
LOG_LEVEL=debug

DB_CONNECTION=sqlite
DB_DATABASE=/your-path/HorizonTest/database/database.sqlite
DB_USERNAME=root
DB_PASSWORD=

BROADCAST_DRIVER=log
CACHE_DRIVER=file
FILESYSTEM_DRIVER=local
QUEUE_CONNECTION=redis
SESSION_DRIVER=file
SESSION_LIFETIME=120

MEMCACHED_HOST=127.0.0.1

REDIS_HOST=127.0.0.1
REDIS_PASSWORD=null
REDIS_PORT=6379

REDIS_QUEUE_HOST=external-redis-ip-here
REDIS_QUEUE_PASSWORD=null
REDIS_QUEUE_PORT=6379

MAIL_MAILER=smtp
MAIL_HOST=mailhog
MAIL_PORT=1025
MAIL_USERNAME=null
MAIL_PASSWORD=null
MAIL_ENCRYPTION=null
MAIL_FROM_ADDRESS=null
MAIL_FROM_NAME="${APP_NAME}"

AWS_ACCESS_KEY_ID=
AWS_SECRET_ACCESS_KEY=
AWS_DEFAULT_REGION=us-east-1
AWS_BUCKET=
AWS_USE_PATH_STYLE_ENDPOINT=false

PUSHER_APP_ID=
PUSHER_APP_KEY=
PUSHER_APP_SECRET=
PUSHER_APP_CLUSTER=mt1

MIX_PUSHER_APP_KEY="${PUSHER_APP_KEY}"
MIX_PUSHER_APP_CLUSTER="${PUSHER_APP_CLUSTER}"
```

**Comment by sp4r74cus** [comment]
I faced the same issue today when trying to separate queues to different connections.

Horizon Version: 5.9.6
Laravel Version: 9.8.1
PHP Version: 8.1.4
Redis Driver: phpredis 5.3.5
Database Driver & Version: MySQL 8

Steps To Reproduce:

1. created a clean Laravel app and installed Horizon.

2. configured new redis connection (Digital Ocean redis database) in `config/database.php`
```
'redis' => [
     ...
     
      'digitalocean' => [
            'url' => env('DO_REDIS_URL'),
        ],
]
```

3. configured a new queue connection that uses the DO server in `config/queue.php`
```
        'digitalocean-redis' => [
            'driver' => 'redis',
            'connection' => 'digitalocean',
            'queue' => env('REDIS_QUEUE', 'default'),
            'retry_after' => 90,
            'block_for' => null,
            'after_commit' => false,
        ],
```

4. configured environments in `config/horizon.php`

```
	'digitalocean' => [
		'connection' => 'digitalocean-redis',
		'queue' => ['digitalocean'],
	]
```

5. created 2 jobs, one dispatched on 'default' queue which is configured to process on the default redis connection (local) , and the second one dispatched on 'digitalocean' queue that is set to process on the digitalocean connection


`.env` 

```
REDIS_HOST=127.0.0.1
REDIS_PASSWORD=null
REDIS_PORT=6379

#REDIS_URL=rediss://default:*******@db-redis-mailcoach-do-user-353183-0.b.db.ondigitalocean.com:25061
DO_REDIS_URL=rediss://default:*******@db-redis-mailcoach-do-user-353183-0.b.db.ondigitalocean.com:25061

```

Results:

'default' queue is processed and 'digitalocean' queue job stays pending.

-----------------------------------

just to be sure all is set correctly, i set the default redis connection to digitalocean and all jobs dispatched and processed.
so i still had 2 connections, and the queues were using both, and all jobs processed as expected.

then i switched connections:
- 'default' redis connection was configured to digitalocean redis server
- 'digitalocean' was now configured to the local redis server

results were now opposite. the job that dispatched on the remote server was processed and the other one one pending.



Hope that helps.

**Comment by driesvints** [comment]
@matt-ostrom-hall-rss I'm not sure if I'll ever be able to reproduce that. It worked for me the last time I tried. I'm not convinced there's an issue but rather a configuration issue. It's best that you try a support channel:

- [Laracasts Forums](https://laracasts.com/discuss)
- [Laravel.io Forums](https://laravel.io/forum)
- [StackOverflow](https://stackoverflow.com/questions/tagged/laravel)
- [Discord](https://discordapp.com/invite/KxwQuKb)
- [Larachat](https://larachat.co)
- [IRC](https://web.libera.chat/?nick=laravelnewbie&channels=#laravel)

**Comment by sp4r74cus** [comment]
So it really was a configuration issue. 

The jobs were dispatched without inConnection() value so they all went into the default one. 

My confusion was that I thought Horizon handles that when I set the connection in the supervisor. But I now understand it only look for dispatched jobs there. 

We learn a new thing every day :)

Thanks for your time replying and linking to those resources 🙏
